File: C++Tokenizer/C++Tokenizer/Tokenizer.cpp
//
//  Tokenizer.cpp
//  tokenizer
//
//  Created by Alex Nagelkerke on 19-09-14.
//  Copyright (c) 2014 Alex Nagelkerke. All rights reserved.
//

#include <stdio.h>
#include <iostream>
#include <algorithm>
#include <functional>
#include "Tokenizer.h"

Tokenizer::Tokenizer(std::string fileLocation, std::list<TokenDefinition> definitions, std::list<TokenPartner> partners)
{
	// set defaults
	lineNumber = 0;
	linePosition = 1;
	level = 1;
	lineRemaining = "";

	tokenDefinitions = definitions;
	tokenPartners = partners;

	file.open(fileLocation);

	nextLine();
}

void Tokenizer::Tokenize()
{
	tokenList = new std::list<Token>();

	while (lineRemaining.length() != 0)
	{
		lineRemaining = trim(lineRemaining);
		bool match = false;

		std::list<TokenDefinition>::iterator tokenDef;
		for (tokenDef = tokenDefinitions.begin(); tokenDef != tokenDefinitions.end(); ++tokenDef)
		{
			TokenDefinition tokenD = *tokenDef;
			int matched = tokenD.matcher->Match(lineRemaining);

			if (matched > 0)
			{
				match = true;

				std::string tokenValue = lineRemaining.substr(0, matched);

				if (tokenValue == "(" || tokenValue == "{" || tokenValue == "[")
					level++;

				Token *partner = nullptr;
				if (tokenD.tokenType == TokenType::CloseCurlyBracket ||
					tokenD.tokenType == TokenType::CloseBracket ||
					tokenD.tokenType == TokenType::Keyword)
					partner = findPartner(tokenValue, level);

				Token *token = new Token(lineNumber, linePosition, level, tokenValue, tokenD.tokenType, partner);

				tokenList->push_back(*token);

				if (tokenValue == ")" || tokenValue == "}" || tokenValue == "]")
					level--;

				linePosition += matched;
				lineRemaining = lineRemaining.substr(matched);

				if (lineRemaining.length() == 0)
					nextLine();

				break;
			}
		}
	}
}

Token* Tokenizer::findPartner(std::string &tokenStr, int level)
{
	Token *token = nullptr;
	std::list<TokenPartner>::iterator tokenPartner;
	for (tokenPartner = tokenPartners.begin(); tokenPartner != tokenPartners.end(); ++tokenPartner)
	{
		TokenPartner tokenP = *tokenPartner;
		if (tokenP.token == tokenStr)
		{
			std::list<Token>::reverse_iterator tokenIt;
			for (tokenIt = tokenList->rbegin(); tokenIt != tokenList->rend(); ++tokenIt)
			{
				token = &(*tokenIt);
				if (token->Value == tokenP.partner && token->Level == level)
				{
					return token;
				}
			}
			return token;
		}
	}

	return token;
}

void Tokenizer::nextLine()
{
	while (std::getline(file, lineRemaining))
	{
		++lineNumber;
		linePosition = 1;

		if (lineRemaining.length() > 0)
			break;
	}
}

std::list<Token>* Tokenizer::GetTokenList()
{
	return tokenList;
}

// From Stackoverflow http://stackoverflow.com/a/217605
// trim from start
std::string Tokenizer::ltrim(std::string &s) {
	s.erase(s.begin(), std::find_if(s.begin(), s.end(), std::not1(std::ptr_fun<int, int>(std::isspace))));
	return s;
}

// trim from end
std::string Tokenizer::rtrim(std::string &s) {
	s.erase(std::find_if(s.rbegin(), s.rend(), std::not1(std::ptr_fun<int, int>(std::isspace))).base(), s.end());
	return s;
}

// trim from both ends
std::string Tokenizer::trim(std::string &s) {
	std::string temp = rtrim(s);
	return ltrim(temp);
}

File: DP_Compiler/DP_tokenizer/Tokenizer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DP_Tokenizer
{
    class Tokenizer
    {
        private readonly TextReader _reader;
        private readonly List<TokenDefinition> _tokenDefinitions;
        private readonly List<TokenPartner> _tokenPartners;

        private TokenList<Token> tokenList;

        private int _lineNumber  = 0;
        private int _linePostion = 1;
        private int _level       = 1;
        private string _lineRemaining;

        public Tokenizer(TextReader reader, List<TokenDefinition> tokenDefinitions, List<TokenPartner> tokenPartners)
        {
            _reader = reader;
            _tokenDefinitions = tokenDefinitions;
            _tokenPartners = tokenPartners;
            nextLine();
        }

        public void Tokenize()
        {
            tokenList = new TokenList<Token>();

            while (_lineRemaining != null)
            {
                _lineRemaining = _lineRemaining.Trim();
                bool match = false;
    
                foreach(TokenDefinition tokenDefinition in _tokenDefinitions)
                {
                    int matched = tokenDefinition.matcher.Match(_lineRemaining);
                    if (matched > 0)
                    {
                        match = true;

                        string tokenValue = _lineRemaining.Substring(0, matched);

                        if (tokenValue == "(" || tokenValue == "{" || tokenValue == "[")
                            _level++;

                        Token partner = null;
                        if (tokenDefinition.tokenType == TokenType.Else || tokenDefinition.tokenType == TokenType.ElseIf ||
                            tokenDefinition.tokenType == TokenType.CloseBracket || tokenDefinition.tokenType == TokenType.While ||
                            tokenDefinition.tokenType == TokenType.CloseCurlyBracket || tokenDefinition.tokenType == TokenType.CloseParenthesis)
                            partner = findPartner(tokenDefinition.tokenType, _level);

                        Token token = new Token(_lineNumber, _linePostion, _level, tokenValue, tokenDefinition.tokenType, partner);
                        tokenList.AddLast(token);

                        // Add token to the found partner
                        if (partner != null)
                            partner.Partner = token;

                        if (tokenValue == ")" || tokenValue == "}" || tokenValue == "]")
                            _level--;

                        _linePostion  += matched; 
                        _lineRemaining = _lineRemaining.Substring(matched);
                        if (_lineRemaining.Length == 0)
                            nextLine();

                        break;
                    }
                }

                if (!match)
                {
                    throw new ParserException("Unexpected character '" + _lineRemaining + "' at line " + _lineNumber + " position" + _linePostion);
                }
            } 
        }

        private Token findPartner(TokenType tokenType, int level)
        {
            Token token = null;

            foreach (TokenPartner tokenPartner in _tokenPartners)
            {

                if (tokenPartner.Token == tokenType)
                {
                    foreach (Token lToken in tokenList)
                    {
                        if (lToken.Type == tokenPartner.Partner && lToken.Level == level && lToken.Partner == null)
                            return lToken;
                    }
                }
            }

            return token;
        }

        private void nextLine()
        {
            do
            {
                _lineRemaining = _reader.ReadLine();
                ++_lineNumber;
                _linePostion = 1;
            }
            while (_lineRemaining != null && _lineRemaining.Length == 0);
        }

        public TokenList<Token> GetTokenList()
        {
            return tokenList;
        }
    }
}
